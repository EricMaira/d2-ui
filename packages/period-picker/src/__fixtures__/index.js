export const periodTypes = {
    Daily: 'Daily',
    Weekly: 'Weekly',
    WeeklyWednesday: 'Weekly Wednesday',
    WeeklyThursday: 'Weekly Thursday',
    WeeklySaturday: 'Weekly Saturday',
    WeeklySunday: 'Weekly Sunday',
    BiWeekly: 'Bi weekly',
    Monthly: 'Monthly',
    BiMonthly: 'Bi-monthly',
    Quarterly: 'Quarterly',
    SixMonthly: 'Six monthly',
    SixMonthlyApril: 'Six monthly starting in April',
    SixMonthlyNov: 'Six monthly starting in November',
    Yearly: 'Yearly',
    FinancialApril: 'Financial year starting in April',
    FinancialJuly: 'Financial year starting in July',
    FinancialOct: 'Financial year starting in October',
    FinancialNov: 'Financial year starting in November',
};

export const getPeriodTypesResponse = {
    periodTypes: [
        {
            frequencyOrder: 1,
            name: 'Daily',
            isoDuration: 'P1D',
            isoFormat: 'yyyyMMdd',
        },
        {
            frequencyOrder: 7,
            name: 'Weekly',
            isoDuration: 'P7D',
            isoFormat: 'yyyyWn',
        },
        {
            frequencyOrder: 7,
            name: 'WeeklyWednesday',
            isoDuration: 'P7D',
            isoFormat: 'yyyyWedWn',
        },
        {
            frequencyOrder: 7,
            name: 'WeeklyThursday',
            isoDuration: 'P7D',
            isoFormat: 'yyyyThuWn',
        },
        {
            frequencyOrder: 7,
            name: 'WeeklySaturday',
            isoDuration: 'P7D',
            isoFormat: 'yyyySatWn',
        },
        {
            frequencyOrder: 7,
            name: 'WeeklySunday',
            isoDuration: 'P7D',
            isoFormat: 'yyyySunWn',
        },
        {
            frequencyOrder: 14,
            name: 'BiWeekly',
            isoDuration: 'P14D',
            isoFormat: 'yyyyBiWn',
        },
        {
            frequencyOrder: 30,
            name: 'Monthly',
            isoDuration: 'P1M',
            isoFormat: 'yyyyMM',
        },
        {
            frequencyOrder: 61,
            name: 'BiMonthly',
            isoDuration: 'P2M',
            isoFormat: 'yyyyMMB',
        },
        {
            frequencyOrder: 91,
            name: 'Quarterly',
            isoDuration: 'P3M',
            isoFormat: 'yyyyQn',
        },
        {
            frequencyOrder: 182,
            name: 'SixMonthly',
            isoDuration: 'P6M',
            isoFormat: 'yyyySn',
        },
        {
            frequencyOrder: 182,
            name: 'SixMonthlyApril',
            isoDuration: 'P6M',
            isoFormat: 'yyyyAprilSn',
        },
        {
            frequencyOrder: 182,
            name: 'SixMonthlyNov',
            isoDuration: 'P6M',
            isoFormat: 'yyyyNovSn',
        },
        {
            frequencyOrder: 365,
            name: 'Yearly',
            isoDuration: 'P1Y',
            isoFormat: 'yyyy',
        },
        {
            frequencyOrder: 365,
            name: 'FinancialApril',
            isoDuration: 'P1Y',
            isoFormat: 'yyyyApril',
        },
        {
            frequencyOrder: 365,
            name: 'FinancialJuly',
            isoDuration: 'P1Y',
            isoFormat: 'yyyyJuly',
        },
        {
            frequencyOrder: 365,
            name: 'FinancialOct',
            isoDuration: 'P1Y',
            isoFormat: 'yyyyOct',
        },
        {
            frequencyOrder: 365,
            name: 'FinancialNov',
            isoDuration: 'P1Y',
            isoFormat: 'yyyyNov',
        },
    ],
};
