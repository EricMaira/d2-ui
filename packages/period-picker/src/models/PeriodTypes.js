import i18n from '@dhis2/d2-i18n';
import {
    BI_MONTH,
    BI_WEEK,
    DAY,
    MONTH,
    QUARTER,
    SIX_MONTH,
    WEEK,
    YEAR,
} from '@dhis2/d2-ui-period-picker/src/models/distinctTypes';
import {
    DailyPeriodType,
    WeeklyPeriodType,
    BiWeeklyPeriodType,
    MonthlyPeriodType,
    BiMonthlyPeriodType,
    QuarterlyPeriodType,
    SixMonthlyPeriodType,
    YearlyPeriodType,
} from '@dhis2/d2-ui-period-picker/src/models/PeriodType';

const fieldLabels = {
    [WEEK]: i18n.t('Week'),
    [SIX_MONTH]: i18n.t('Six month period'),
    [YEAR]: i18n.t('Year'),
};

export class PeriodTypes {
    constructor(locale) {
        // const locale = d2.currentUser.userSettings.get('keyUiLocale');
        this.types = new Map([
            [
                'Daily',
                new DailyPeriodType({
                    type: DAY,
                    typeLabel: i18n.t('Daily'),
                    fieldLabel: i18n.t('Day'),
                    requiredSiblingFields: [YEAR, MONTH],
                    locale,
                }),
            ],
            [
                'Weekly',
                new WeeklyPeriodType({
                    type: WEEK,
                    typeLabel: i18n.t('Weekly'),
                    fieldLabel: fieldLabels[WEEK],
                    infix: 'W',
                }),
            ],
            [
                'WeeklyWednesday',
                new WeeklyPeriodType({
                    type: WEEK,
                    typeLabel: i18n.t('Weekly Wednesday'),
                    fieldLabel: fieldLabels[WEEK],
                    infix: 'WedW',
                }),
            ],
            [
                'WeeklyThursday',
                new WeeklyPeriodType({
                    type: WEEK,
                    typeLabel: i18n.t('Weekly Thursday'),
                    fieldLabel: fieldLabels[WEEK],
                    infix: 'ThuW',
                }),
            ],
            [
                'WeeklySaturday',
                new WeeklyPeriodType({
                    type: WEEK,
                    typeLabel: i18n.t('Weekly Saturday'),
                    fieldLabel: fieldLabels[WEEK],
                    infix: 'SatW',
                }),
            ],
            [
                'WeeklySunday',
                new WeeklyPeriodType({
                    type: WEEK,
                    typeLabel: i18n.t('Weekly Sunday'),
                    fieldLabel: fieldLabels[WEEK],
                    infix: 'SunW',
                }),
            ],
            [
                'BiWeekly',
                new BiWeeklyPeriodType({
                    type: BI_WEEK,
                    typeLabel: i18n.t('Bi weekly'),
                    fieldLabel: i18n.t('Bi Week'),
                    infix: 'BiW',
                }),
            ],
            [
                'Monthly',
                new MonthlyPeriodType({
                    type: MONTH,
                    typeLabel: i18n.t('Monthly'),
                    fieldLabel: i18n.t('Month'),
                }),
            ],
            [
                'BiMonthly',
                new BiMonthlyPeriodType({
                    type: BI_MONTH,
                    typeLabel: i18n.t('Bi-monthly'),
                    fieldLabel: i18n.t('Bi-month'),
                    suffix: 'B',
                }),
            ],
            [
                'Quarterly',
                new QuarterlyPeriodType({
                    type: QUARTER,
                    typeLabel: i18n.t('Quarterly'),
                    fieldLabel: i18n.t('Quarter'),
                    infix: 'Q',
                }),
            ],
            [
                'SixMonthly',
                new SixMonthlyPeriodType({
                    type: SIX_MONTH,
                    typeLabel: i18n.t('Six monthly'),
                    fieldLabel: fieldLabels[SIX_MONTH],
                    infix: 'S',
                }),
            ],
            [
                'SixMonthlyApril',
                new SixMonthlyPeriodType({
                    type: SIX_MONTH,
                    typeLabel: i18n.t('Six monthly starting in April'),
                    fieldLabel: fieldLabels[SIX_MONTH],
                    infix: 'AprilS',
                }),
            ],
            [
                'SixMonthlyNov',
                new SixMonthlyPeriodType({
                    type: SIX_MONTH,
                    typeLabel: i18n.t('Six monthly starting in November'),
                    fieldLabel: fieldLabels[SIX_MONTH],
                    infix: 'NovS',
                }),
            ],
            [
                'Yearly',
                new YearlyPeriodType({
                    type: YEAR,
                    typeLabel: i18n.t('Yearly'),
                    fieldLabel: fieldLabels[YEAR],
                }),
            ],
            [
                'FinancialApril',
                new YearlyPeriodType({
                    type: YEAR,
                    typeLabel: i18n.t('Financial year starting in April'),
                    fieldLabel: fieldLabels[YEAR],
                    suffix: 'April',
                }),
            ],
            [
                'FinancialJuly',
                new YearlyPeriodType({
                    type: YEAR,
                    typeLabel: i18n.t('Financial year starting in July'),
                    fieldLabel: fieldLabels[YEAR],
                    suffix: 'July',
                }),
            ],
            [
                'FinancialOct',
                new YearlyPeriodType({
                    type: YEAR,
                    typeLabel: i18n.t('Financial year starting in October'),
                    fieldLabel: fieldLabels[YEAR],
                    suffix: 'Oct',
                }),
            ],
            [
                'FinancialNov',
                new YearlyPeriodType({
                    type: YEAR,
                    typeLabel: i18n.t('Financial year starting in November'),
                    fieldLabel: fieldLabels[YEAR],
                    suffix: 'Nov',
                }),
            ],
        ]);
    }

    getPeriodTypes() {
        const periodTypes = [];
        this.types.forEach((type, key) => {
            periodTypes.push({ label: type.getTypeLabel(), value: key });
        });
        return periodTypes;
    }

    getPeriodType(name) {
        return this.types.get(name);
    }

    getFields(state) {
        if (!state.periodType) {
            return {};
        }
        return this.types.get(state.periodType).getFields(state);
    }

    getFieldUpdateObject({ type, id, startDate }) {
        return this.types.get(type).getFieldUpdateObject(id, startDate);
    }

    findActivePeriodId(state, fieldKeyToSkip) {
        const fields = this.getFields(state);
        let option;

        for (let fieldKey in fields) {
            if (fieldKey !== fieldKeyToSkip) {
                const field = fields[fieldKey];
                const fieldValue = state[field.name];
                option = field.options.find(
                    option => option.id && option.value === fieldValue
                );
                // Stop looping once found
                if (option) {
                    break;
                }
            }
        }
        return option ? option.id : null;
    }
}

export default PeriodTypes;
