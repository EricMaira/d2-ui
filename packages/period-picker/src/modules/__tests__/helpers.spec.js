import {
    getYears,
    getMonths,
    createSequence,
    zeroPad,
    asInts,
    getMonthFromId,
    getYearFromId,
    checkForUnsupportedPeriodTypes,
} from '../helpers';
import { getPeriodTypesResponse } from '../../__fixtures__';
import PeriodTypes from '../PeriodTypes';

describe('periodTypes helpers', () => {
    describe('getYears', () => {
        const MOCK_YEAR = 2011;
        beforeAll(() => {
            jest.spyOn(Date.prototype, 'getFullYear').mockImplementation(
                () => MOCK_YEAR
            );
        });

        afterAll(() => {
            jest.restoreAllMocks();
        });

        it('Returns the correct range of years', () => {
            const years = getYears({}).options;
            const startYear = (MOCK_YEAR - 4).toString();
            const endYear = (MOCK_YEAR + 3).toString();

            expect(years.length).toEqual(8);
            expect(years[0]).toEqual({ value: startYear, label: startYear });
            expect(years[7]).toEqual({ value: endYear, label: endYear });
        });

        it('Works with a negative offset', () => {
            const state = {
                yearOffset: -1,
            };
            const years = getYears(state).options;
            const startYear = (MOCK_YEAR - 4 - 8).toString();
            const endYear = (MOCK_YEAR + 3 - 8).toString();

            expect(years.length).toEqual(8);
            expect(years[0]).toEqual({ value: startYear, label: startYear });
            expect(years[7]).toEqual({ value: endYear, label: endYear });
        });
        it('Works with a positive offset', () => {
            const state = {
                yearOffset: 6,
            };
            const years = getYears(state).options;
            const startYear = (MOCK_YEAR - 4 + 8 * 6).toString();
            const endYear = (MOCK_YEAR + 3 + 8 * 6).toString();

            expect(years.length).toEqual(8);
            expect(years[0]).toEqual({ value: startYear, label: startYear });
            expect(years[7]).toEqual({ value: endYear, label: endYear });
        });
    });

    describe('getMonths', () => {
        it('Produces the correct output', () => {
            expect(getMonths()).toMatchSnapshot();
        });
    });

    describe('createSequence', () => {
        it('Produces an array of strings starting at "1"', () => {
            const sequence = createSequence(10);
            expect(sequence.length).toEqual(10);
            expect(sequence[0]).toEqual('1');
            expect(sequence[9]).toEqual('10');
        });
        it('Can zeroPad the strings', () => {
            const sequence = createSequence(10, true);
            expect(sequence.length).toEqual(10);
            expect(sequence[0]).toEqual('01');
            expect(sequence[9]).toEqual('10');
        });
        it('Can start at another number', () => {
            const sequence = createSequence(10, false, 91);
            expect(sequence.length).toEqual(10);
            expect(sequence[0]).toEqual('91');
            expect(sequence[9]).toEqual('100');
        });
    });

    describe('zeroPad', () => {
        it('Returns a string', () => {
            expect(typeof zeroPad(1)).toEqual('string');
        });
        it('Adds leading zero when needed', () => {
            expect(zeroPad(1)).toEqual('01');
        });
        it('Does not add leading zero when the number > 9', () => {
            expect(zeroPad(10)).toEqual('10');
        });
    });

    describe('asInts', () => {
        it('Returns a number representation of strings', () => {
            const state = {
                year: '2011',
                month: '01',
                week: '1',
                day: '04',
            };
            const propKeys = ['year', 'month', 'week', 'day'];
            expect(asInts(state, propKeys)).toEqual([2011, 1, 1, 4]);
        });
    });
    describe('getMonthFromId', () => {
        it('Returns the correct month string', () => {
            expect(getMonthFromId('201104')).toEqual('04');
        });
    });
    describe('getYearFromId', () => {
        it('Returns the correct year string', () => {
            expect(getYearFromId('20110401')).toEqual('2011');
            expect(getYearFromId('2011WedW6')).toEqual('2011');
        });
    });
});

describe('checkForUnsupportedPeriodTypes', () => {
    const mockedWarn = jest.fn();
    const ORIGINAL_CONSOLE_WARN = console.warn;
    const ORIGINAL_NODE_ENV = process.env.NODE_ENV;
    const unsupportedTypes = [{ name: 'Faker' }, { name: 'Another faker' }];
    const periodTypesListWithUnSupported = [
        ...getPeriodTypesResponse.periodTypes,
        ...unsupportedTypes,
    ];
    const mockedResponse = {
        periodTypes: periodTypesListWithUnSupported,
    };
    const mockedGet = jest.fn(() => Promise.resolve(mockedResponse));
    const mockedD2 = {
        Api: {
            getApi: () => ({ get: mockedGet }),
        },
    };
    const periodTypes = new PeriodTypes('en');
    const getPeriodType = periodTypes.getPeriodType.bind(periodTypes);

    beforeAll(() => {
        console.warn = mockedWarn;
    });

    beforeEach(() => {
        jest.resetModules();
        mockedWarn.mockClear();
    });

    afterAll(() => {
        console.warn = ORIGINAL_CONSOLE_WARN;
        process.env.NODE_ENV = ORIGINAL_NODE_ENV;
    });

    describe('If NODE_ENV is "development"', () => {
        it('Will produce a console warning if unsupported types are found', () => {
            process.env.NODE_ENV = 'development';
            const message = [
                'WARNING: Unsupported period type(s) detected',
                'The PeriodPicker component needs to be updated to support the following period type(s):',
                'Faker, Another faker',
            ].join('\n');

            return checkForUnsupportedPeriodTypes(mockedD2, getPeriodType).then(
                () => {
                    expect(mockedWarn).toHaveBeenCalledTimes(1);
                    expect(mockedWarn).toHaveBeenCalledWith(message);
                }
            );
        });

        it('Will NOT produce a console warning if all types are supported', () => {
            process.env.NODE_ENV = 'development';
            mockedResponse.periodTypes = getPeriodTypesResponse.periodTypes;
            return checkForUnsupportedPeriodTypes(mockedD2, getPeriodType).then(
                () => {
                    expect(mockedWarn).toHaveBeenCalledTimes(0);
                }
            );
        });
    });

    describe('If NODE_ENV is "production"', () => {
        it('Will NOT produce a console warning, even if unsupported types are found', () => {
            process.env.NODE_ENV = 'production';
            mockedResponse.periodTypes = periodTypesListWithUnSupported;
            checkForUnsupportedPeriodTypes(periodTypesListWithUnSupported);
            return checkForUnsupportedPeriodTypes(mockedD2, getPeriodType).then(
                () => {
                    expect(mockedWarn).toHaveBeenCalledTimes(0);
                }
            );
        });
    });
});
