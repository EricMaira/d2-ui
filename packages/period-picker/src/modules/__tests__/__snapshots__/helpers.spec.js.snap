// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`periodTypes helpers getMonths Produces the correct output 1`] = `
Object {
  "label": "Month",
  "name": "month",
  "options": Array [
    Object {
      "label": "January",
      "value": "01",
    },
    Object {
      "label": "February",
      "value": "02",
    },
    Object {
      "label": "March",
      "value": "03",
    },
    Object {
      "label": "April",
      "value": "04",
    },
    Object {
      "label": "May",
      "value": "05",
    },
    Object {
      "label": "June",
      "value": "06",
    },
    Object {
      "label": "July",
      "value": "07",
    },
    Object {
      "label": "August",
      "value": "08",
    },
    Object {
      "label": "September",
      "value": "09",
    },
    Object {
      "label": "October",
      "value": "10",
    },
    Object {
      "label": "November",
      "value": "11",
    },
    Object {
      "label": "December",
      "value": "12",
    },
  ],
}
`;
